_posts/2021-02-09-message-passing.md

```diff
@@ -0,0 +1,188 @@
+        %% Undo monitor/2 and remove any 'DOWN' messages that might have
+        %% arrived in the meantime.
+        erlang:demonitor(Mref, [flush]),
+        {ok, Response};
+    {'DOWN', Mref, _, _, Reason} ->
+        {error, Reason}
+end
+```
+
+Since dead processes cannot send messages we know that the response must come
+before any eventual 'DOWN' message, but without a guaranteed order the 'DOWN'
+message could arrive before the response and we'd have no idea whether a
+response was coming or not, which would be very annoying to deal with.
+
+This saves us quite a bit of hassle and doesn't come at much of a cost, but
+the guarantees stop there. If more than one process sends messages to a common
+process, they can arrive in any order even when you "know" that one of the
+messages was sent first. For example, this sequence of events is legal and
+entirely possible:
+
+1. `A` sends message `1` to `B`
+1. `A` sends message `2` to `C`
+1. `C`, in response to message `2`, sends message `3` to `B`
+1. `B` receives message `3`
+1. `B` receives message `1`
+
+Luckily, global orders are rarely needed and are trivial to impose yourself:
+just let all involved parties synchronize with a common process.
+
+### Sending messages
+
+Sending a message is straightforward: we try to find the process associated
+with the process identifier, and if one exists we insert the message into its
+signal queue.
+
+Messages are always copied before being inserted into the queue. As wasteful as
+this may sound it greatly reduces garbage collection (GC) latency as the GC
+never has to look beyond a single process, which is more important than sheer
+throughput for the kind of soft-realtime systems Erlang is designed to build.
+
+By default, messages are copied directly into the receiving process' heap but
+when this isn't possible (or desired -- see the [message_queue_data] flag) we
+allocate the message outside of the heap instead.
+
+Memory allocation makes such "off-heap" messages slightly more expensive but
+they're very neat for processes that receive a ton of messages. We don't need
+to interact with the receiver when copying the message -- only when adding it
+to the queue -- and since the only way a process can see a message is by
+matching them in a `receive` expression, the GC doesn't need to consider
+unmatched messages which further reduces latency.
+
+### Receiving messages
+
+One might guess that processes receive messages through `receive` expressions,
+but `receive` is a bit of a misnomer. As with all other signals the process
+continuously handles them in the background, moving received messages from the
+_signal_ queue to the _message_ queue.
+
+`receive` searches for matching messages in the message queue (in the order
+they arrived), or waits for new messages if none were found. Acting on the
+message rather than signal queue means it doesn't have to contend with
+processes that send messages, which greatly increases performance.
```

Comment:
I found this sentence difficult to understand. Maybe you can expand on how `receive` is "Acting on the message rather than [the] signal queue". Is what you mean that `receive` is acting on the internal queue rather than the external one that other processes can write to?

Reply:
It searches through the message queue, not the signal queue. I'll change "acting" to "searching" instead.